### PR TITLE
feat(issue-platform): Send issue occurrences to eventstream when saving them.

### LINKED ISSUE #42054: Ingestion: Pass event and issueoccurrence to eventstream
Ideally we can just pass a `GroupEvent` to eventstream here with the `IssueOccurrence` included inside of it. We currently have some safety checks that ensure that `GroupEvents` are never passed here. These should remain true when dealing with `transactions` and `errors` for now, but we should allow them to be passed for `platform` events.

### PR DESCRIPTION
**Pull request by wedamija**
This hooks up the work done in https://github.com/getsentry/sentry/pull/42195 to the occurrence saving process. Tests will fail until we have a dataset available in snuba.

### CODE REVIEW

File: src/sentry/issues/ingest.py
Comment (ceorourke) [inline src/sentry/issues/ingest.py:200]: Is `event_group` a `GroupEvent`? 😵 Might be less mind bendy if you call this `group_event` imo.

File: src/sentry/issues/ingest.py
Comment (ceorourke) [inline src/sentry/issues/ingest.py:209]: Is this comment about the `received_timestamp`? What do we do for performance issues?

File: tests/sentry/issues/test_ingest.py
Comment (ceorourke) [inline tests/sentry/issues/test_ingest.py:36]: Is this blocked on https://github.com/getsentry/snuba/pull/3468?

File: tests/sentry/issues/test_ingest.py
Comment (ceorourke) [inline tests/sentry/issues/test_ingest.py:186]: What's a platform event? Like the `load_data` we use in other tests?

**Comment by barkbarkimashark** [comment]
Ugh, I was just testing this locally against the new dataset and we might have to introduce a  `EVENT_PROTOCOL_VERSION` mapping from `EventStreamEventType -> version`
https://github.com/getsentry/sentry/blob/a3a9cd1cb0d3e14a20d88e3fd13c5c256c60eca7/src/sentry/eventstream/snuba.py#L398

It looks like errors and transactions are both using version=2. I think it makes sense to separate it out so `errors/transactions version == 2`, and `issue_platform == 1`.

File: tests/sentry/issues/test_ingest.py
Comment (wedamija) [reply tests/sentry/issues/test_ingest.py:186]: It's an event that we received via the ingestion consumer that @rjo100 is working on. We might rename them to `generic` events based on the other pr, but we'll see. Basically just needs to be not an error or transaction event.

**Comment by wedamija** [comment]
> Ugh, I was just testing this locally against the new dataset and we might have to introduce a `EVENT_PROTOCOL_VERSION` mapping from `EventStreamEventType -> version`
> 
> https://github.com/getsentry/sentry/blob/a3a9cd1cb0d3e14a20d88e3fd13c5c256c60eca7/src/sentry/eventstream/snuba.py#L398
> 
> It looks like errors and transactions are both using version=2. I think it makes sense to separate it out so `errors/transactions version == 2`, and `issue_platform == 1`.

Is there any reason we can't use version 2 as well? What are the differences?

File: tests/sentry/issues/test_ingest.py
Comment (wedamija) [reply tests/sentry/issues/test_ingest.py:36]: Yeah, once that is merged we'll have the dataset available and can write to it via eventstream and query as well.

File: src/sentry/issues/ingest.py
Comment (wedamija) [reply src/sentry/issues/ingest.py]: We use `received` from the event and then if unavailable the event datetime, so I'm just going to do the same here

File: src/sentry/issues/ingest.py
Comment (wedamija) [reply src/sentry/issues/ingest.py]: I feel like my brain just randomly scrambled the order here 🤯

**Comment by rjo100** [comment]
lgtm once we get the upstream details hammered out

**Comment by barkbarkimashark** [comment]
> > Ugh, I was just testing this locally against the new dataset and we might have to introduce a `EVENT_PROTOCOL_VERSION` mapping from `EventStreamEventType -> version`
> > https://github.com/getsentry/sentry/blob/a3a9cd1cb0d3e14a20d88e3fd13c5c256c60eca7/src/sentry/eventstream/snuba.py#L398
> > 
> > It looks like errors and transactions are both using version=2. I think it makes sense to separate it out so `errors/transactions version == 2`, and `issue_platform == 1`.
> 
> Is there any reason we can't use version 2 as well? What are the differences?

We can use version 2 for now, but any bump in the version will affect all three processors (error/transaction/searchissue).

**Comment by wedamija** [comment]
> > > Ugh, I was just testing this locally against the new dataset and we might have to introduce a `EVENT_PROTOCOL_VERSION` mapping from `EventStreamEventType -> version`
> > > https://github.com/getsentry/sentry/blob/a3a9cd1cb0d3e14a20d88e3fd13c5c256c60eca7/src/sentry/eventstream/snuba.py#L398
> > > 
> > > It looks like errors and transactions are both using version=2. I think it makes sense to separate it out so `errors/transactions version == 2`, and `issue_platform == 1`.
> > 
> > 
> > Is there any reason we can't use version 2 as well? What are the differences?
> 
> We can use version 2 for now, but any bump in the version will affect all three processors (error/transaction/searchissue).

Tbh we should probably keep the protocol for all of these the same, they're all doing roughly the same thing anyway.

**Comment by barkbarkimashark** [comment]
> > > > Ugh, I was just testing this locally against the new dataset and we might have to introduce a `EVENT_PROTOCOL_VERSION` mapping from `EventStreamEventType -> version`
> > > > https://github.com/getsentry/sentry/blob/a3a9cd1cb0d3e14a20d88e3fd13c5c256c60eca7/src/sentry/eventstream/snuba.py#L398
> > > > 
> > > > It looks like errors and transactions are both using version=2. I think it makes sense to separate it out so `errors/transactions version == 2`, and `issue_platform == 1`.
> > > 
> > > 
> > > Is there any reason we can't use version 2 as well? What are the differences?
> > 
> > 
> > We can use version 2 for now, but any bump in the version will affect all three processors (error/transaction/searchissue).
> 
> Tbh we should probably keep the protocol for all of these the same, they're all doing roughly the same thing anyway.

I don't mind if we share the same protocol version, especially when its not causing any conflicts. We can always change it if needed.

File: src/sentry/eventstream/base.py
Comment (rjo100) [inline src/sentry/eventstream/base.py:221]: ideally long term we'd just wrap all this in `get_event_type()`

File: src/sentry/eventstream/kafka/backend.py
Comment (rjo100) [inline src/sentry/eventstream/kafka/backend.py:38]: worth renaming this variable for clarity?

File: src/sentry/eventstream/kafka/backend.py
Comment (rjo100) [inline src/sentry/eventstream/kafka/backend.py:136]: worth turning into an `or` for one logical eval? just a style comment not opinionated about it

File: src/sentry/eventstream/kafka/backend.py
Comment (rjo100) [inline src/sentry/eventstream/kafka/backend.py:178]: do we still need this method?

File: tests/sentry/issues/test_ingest.py
Comment (rjo100) [inline tests/sentry/issues/test_ingest.py:187]: do you want me to add a helper method to grab a generic ingestion event?

**Review by rjo100 — COMMENTED** [review]
just a couple style q's & comments

File: src/sentry/eventstream/base.py
Comment (wedamija) [reply src/sentry/eventstream/base.py:221]: Do you mean in `event.get_event_type` or somewhere else?

File: tests/sentry/issues/test_ingest.py
Comment (wedamija) [reply tests/sentry/issues/test_ingest.py:187]: Yeah I think that'd be helpful. We can just add a fixture [here](https://github.com/getsentry/sentry/blob/6c19ea2ebec191f175456bfdeafd8c427883ad14/src/sentry/data/samples/transaction-n-plus-one.json#L1) and load with `load_data`

Comment (rjo100) [reply tests/sentry/issues/test_ingest.py:187]: will add this in a separate PR

**Review by rjo100 — APPROVED** [review]

**Pull request merged**
